File: windows/Services/WindowManager.cs
using System.Diagnostics.CodeAnalysis;
using System.Windows;
using Microsoft.Extensions.Logging;
using Starsky.Desktop.Models;
using Starsky.Desktop.Windows;

namespace Starsky.Desktop.Services;

[SuppressMessage("Sonar",
	"S107: Methods should not have too many parameters",
	Justification = "Needed")]
public class WindowManager(
	SettingsService settings,
	RoutePersistenceService routes,
	NavigationService navigation,
	WebViewEnvironmentService webViewEnv,
	FileDownloadService fileDownload,
	FileWatcherService watcher,
	UpdateService updateService,
	ILogger<WindowManager> logger,
	IMountWatcherService? mountWatcherService = null)
{
	private readonly List<MainWindow> _mainWindows = [];
    private int? _localPort;

    public bool IsTerminating { get; private set; }

    // True when this is the only window left; used by MainWindow_Closing to decide
    // whether to save itself (last window = app is exiting, keep the state).
    internal bool IsLastOpenWindow => _mainWindows.Count == 1;

    public void SetLocalPort(int port) => _localPort = port;

    internal static SavedWindowState ResolveGeometry(
        SavedWindowState? geometry, int offset, ILogger? logger = null)
    {
        if (geometry != null && IsOnScreen(geometry))
        {
	        return geometry;
        }

        if (geometry != null)
        {
	        logger?.LogWarning("Saved window geometry is off-screen; resetting to default position");
        }

        return new SavedWindowState
        {
            Left = 100 + offset,
            Top = 100 + offset,
            Width = 1200,
            Height = 800,
            Route = geometry?.Route ?? "?f=/"
        };
    }

    public void OpenMainWindow(string? route, SavedWindowState? geometry)
    {
        var baseUrl = navigation.GetEffectiveBaseUrl(_localPort);
        var offset = _mainWindows.Count * 24;
        var state = ResolveGeometry(geometry, offset, logger);

        var window = new MainWindow(new MainWindowOptions
        {
            Settings = settings,
            WebViewEnv = webViewEnv,
            FileDownload = fileDownload,
            Watcher = watcher,
            WindowManager = this,
            Logger = logger,
            BaseUrl = baseUrl,
            InitialRoute = route ?? "?f=/",
            Geometry = state,
            UpdateService = updateService,
            MountWatcherService = mountWatcherService
        });

        _mainWindows.Add(window);
        window.Closed += (_, _) =>
        {
            _mainWindows.Remove(window);
            if (_mainWindows.Count == 0 && !IsTerminating && Application.Current != null)
            {
	            Application.Current.Shutdown();
            }
        };

        window.Show();
    }

    public void RestoreWindows()
    {
        var saved = routes.GetRoutes();
        if (saved.Count == 0)
        {
            OpenMainWindow(null, null);
            return;
        }

        foreach (var state in saved)
        {
	        OpenMainWindow(state.Route, state);
        }
    }

    private List<SavedWindowState> CollectStates(MainWindow? exclude = null) =>
        _mainWindows
            .Where(w => w != exclude)
            .Select(w => w.GetCurrentState())
            .ToList();

    internal void PersistCurrentState(MainWindow? exclude = null) =>
        routes.SaveAll(CollectStates(exclude));

    public void CloseAll(bool saveState = true)
    {
        IsTerminating = true;
        // Skip when _mainWindows is already empty: the last window saved its own
        // state in MainWindow_Closing before triggering Shutdown(), so overwriting
        // with an empty list here would discard it.
        if (saveState && _mainWindows.Count > 0)
        {
            routes.SaveAll(CollectStates());
        }

        foreach (var w in _mainWindows.ToList())
        {
            try { w.Close(); } catch { /* best-effort */ }
        }
        _mainWindows.Clear();
    }

    public void ReopenAll()
    {
        CloseAll(saveState: false);
        routes.ClearAll();
        IsTerminating = false;
        OpenMainWindow(null, null);
    }

    public void ReloadAll()
    {
        foreach (var w in _mainWindows)
        {
	        w.Reload();
        }
    }

    // Maximized windows are always valid — WPF snaps them to the nearest screen.
    // For normal windows, require that a usable strip of the title bar (≥100 px wide,
    // ≥1 px of height) intersects the virtual desktop so the user can grab and move it.
    internal static bool IsOnScreen(SavedWindowState state)
    {
        if (state.IsMaximized)
        {
	        return true;
        }

        if (state.Width < 200 || state.Height < 100)
        {
	        return false;
        }

        var vLeft   = SystemParameters.VirtualScreenLeft;
        var vTop    = SystemParameters.VirtualScreenTop;
        var vRight  = vLeft + SystemParameters.VirtualScreenWidth;
        var vBottom = vTop  + SystemParameters.VirtualScreenHeight;

        // Window right edge must be far enough right, and left edge far enough left,
        // that at least 100 px of the title bar is reachable.
        const double minTitleBarVisible = 100;
        return state.Left + state.Width  > vLeft  + minTitleBarVisible
            && state.Left                < vRight - minTitleBarVisible
            && state.Top                 < vBottom
            && state.Top + 30            > vTop;
    }
}

File: mac/starsky/WindowManager.swift
import AppKit

class WindowManager {
    private var windows: [MainWindowController] = []
    private var localPort: Int = 0
    private var isReopening = false
    private let settingsService: SettingsService
    private let routePersistenceService: RoutePersistenceService
    private let navigationService: NavigationService
    private let fileDownloadService: FileDownloadService
    private let fileLogger: DailyFileLogger

    init(
        settingsService: SettingsService,
        routePersistenceService: RoutePersistenceService,
        navigationService: NavigationService,
        fileDownloadService: FileDownloadService,
        fileLogger: DailyFileLogger
    ) {
        self.settingsService = settingsService
        self.routePersistenceService = routePersistenceService
        self.navigationService = navigationService
        self.fileDownloadService = fileDownloadService
        self.fileLogger = fileLogger
    }

    func setLocalPort(_ port: Int) {
        localPort = port
    }

    @MainActor
    func openMainWindow(route: String? = nil, geometry: SavedWindowState? = nil) {
        let index = windows.count
        let baseUrl = navigationService.getEffectiveBaseUrl(localPort: localPort)
        let startUrl = navigationService.buildStartUrl(baseUrl: baseUrl, route: route)
        let options = MainWindowOptions(
            index: index,
            startUrl: startUrl,
            baseUrl: baseUrl,
            mode: settingsService.current.mode,
            geometry: geometry,
            navigationService: navigationService,
            routePersistenceService: routePersistenceService,
            fileDownloadService: fileDownloadService,
            windowManager: self,
            fileLogger: fileLogger
        )
        let controller = MainWindowController(options: options)
        windows.append(controller)

        let cascadeOffset = CGFloat(index) * 24
        let resolved = WindowManager.resolveGeometry(geometry, offset: cascadeOffset)

        controller.window?.setFrame(
            NSRect(x: resolved.x, y: resolved.y, width: resolved.width, height: resolved.height),
            display: false
        )
        if resolved.isMaximized {
            controller.window?.zoom(nil)
        }
        controller.showWindow(nil)
    }

    // Maximized windows are always valid — macOS snaps them to the nearest screen.
    // For normal windows, require that at least 100 px of the title bar is reachable
    // on any connected screen so the user can grab and move the window.
    static func isOnScreen(_ state: SavedWindowState) -> Bool {
        if state.isMaximized { return true }
        if state.width < 200 || state.height < 100 { return false }

        let minVisible: CGFloat = 100
        let winLeft   = CGFloat(state.x)
        let winTop    = CGFloat(state.y)
        let winRight  = winLeft + CGFloat(state.width)
        let winBottom = winTop  + CGFloat(state.height)

        return NSScreen.screens.contains { screen in
            let f = screen.frame
            return winRight  > f.minX + minVisible
                && winLeft   < f.maxX - minVisible
                && winBottom > f.minY
                && winTop    < f.maxY
        }
    }

    static func resolveGeometry(_ geometry: SavedWindowState?, offset: CGFloat) -> SavedWindowState {
        if let g = geometry, isOnScreen(g) {
            var resolved = g
            resolved.x += Double(offset)
            return resolved
        }
        return SavedWindowState(
            route: geometry?.route ?? "?f=/",
            x: 100 + Double(offset),
            y: 100,
            width: 1200,
            height: 800,
            isMaximized: false
        )
    }

    @MainActor
    func restoreWindows() {
        let routes = routePersistenceService.getRoutes()
        if routes.isEmpty {
            openMainWindow()
        } else {
            for state in routes {
                openMainWindow(route: state.route, geometry: state)
            }
        }
    }

    @MainActor
    func closeAll() {
        for controller in windows {
            controller.window?.close()
        }
        windows.removeAll()
    }

    @MainActor
    func openMainWindow(route: String?) {
        openMainWindow(route: route, geometry: nil)
    }

    @MainActor
    func reopenAll() {
        isReopening = true
        routePersistenceService.clearAll()
        closeAll()
        openMainWindow()
        isReopening = false
    }

    @MainActor
    func reloadAll() {
        for controller in windows {
            controller.reload()
        }
    }

    @MainActor
    func remove(controller: MainWindowController) {
        windows.removeAll { $0 === controller }
        if windows.isEmpty && !isReopening {
            NSApplication.shared.terminate(nil)
        }
    }
}

